lib/ansible/plugins/connection/ssh.py

```diff
@@ -309,8 +309,49 @@ def _examine_output(self, source, state, b_chunk, sudoable):
 
         return b''.join(output), remainder
 
+
     def _run(self, cmd, in_data, sudoable=True):
         '''
+        Will retry execution if ssh were unable to connect or timeout
+        '''
+        remaining_tries = int(C.ANSIBLE_SSH_RETRIES) + 1
+        cmd_summary = ' '.join(cmd)
+        for attempt in range(remaining_tries):
+            try:
+                return_tuple = self._run_once(cmd=cmd, in_data=in_data, sudoable=sudoable)
+                # 0 = success
+                # 1-254 = remote command return code
+                # 255 = failure from the ssh command itself
+                if return_tuple[0] != 255:
+                    break
+                else:
+                    raise AnsibleConnectionFailure("Failed to connect to the host via ssh.")
+            except (AnsibleConnectionFailure, AnsibleConnectionTimeout, Exception) as e:
```

Comment:
My bad, I ported that change from Ansible 2.1 and forgot to change Timeout error. That should be raised within _run_once:

```
       while True:
            rfd, wfd, efd = select.select(rpipes, [], [], timeout)

            # We pay attention to timeouts only while negotiating a prompt.

            if not rfd:
                if state <= states.index('awaiting_escalation'):
                    # If the process has already exited, then it's not really a
                    # timeout; we'll let the normal error handling deal with it.
                    if p.poll() is not None:
                        break
                    self._terminate_process(p)
                    raise AnsibleConnectionTimeout('Timeout (%ds) waiting for privilege escalation prompt: %s' % (timeout, to_native(b_stdout)))
```
I'll try to update pull request